ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/DeprecatedStatementsCheckerTest.java

```diff
@@ -0,0 +1,269 @@
+import java.util.Optional;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DeprecatedStatementsCheckerTest {
+  private final KsqlParser STATEMENT_PARSER = new DefaultKsqlParser();
+
+  private final SourceName STREAM_1 = SourceName.of("STREAM_1");
+  private final SourceName STREAM_2 = SourceName.of("STREAM_2");
+  private final SourceName STREAM_3 = SourceName.of("STREAM_3");
+  private final SourceName TABLE_1 = SourceName.of("TABLE_1");
+
+  private DeprecatedStatementsChecker statementsChecker;
+
+  @Mock
+  private MetaStore metaStore;
+
+  @Before
+  public void setup() {
+    statementsChecker = new DeprecatedStatementsChecker(metaStore);
+
+    final DataSource streamSource1 = newKsqlStream(STREAM_1);
+    when(metaStore.getSource(STREAM_1)).thenReturn(streamSource1);
+
+    final DataSource streamSource2 = newKsqlStream(STREAM_2);
+    when(metaStore.getSource(STREAM_2)).thenReturn(streamSource2);
+
+    final DataSource streamSource3 = newKsqlStream(STREAM_3);
+    when(metaStore.getSource(STREAM_3)).thenReturn(streamSource3);
+
+    final DataSource tableSource1 = newKsqlTable(TABLE_1);
+    when(metaStore.getSource(TABLE_1)).thenReturn(tableSource1);
+  }
+
+  @Test
+  public void shouldDeprecateStreamStreamJoinsWithNoGrace() {
+    for (final JoinType joinType : JoinType.values()) {
+      checkDeprecatedStatement(String.format(
+          "SELECT * FROM %s AS l "
+              + "%s JOIN %s AS r WITHIN 1 SECOND ON l.K = r.K;",
+          STREAM_1.text(),
+          (joinType == JoinType.OUTER) ? "FULL OUTER" : joinType,
+          STREAM_2.text()
+      ));
+    }
+  }
+
+  @Test
+  public void shouldDeprecateStreamStreamJoinsWithNoGraceOnMultiStreamStreamJoinOrder1() {
+    for (final JoinType joinType : JoinType.values()) {
+      checkDeprecatedStatement(String.format(
+          "SELECT * FROM %s AS l "
+              + "%s JOIN %s AS m WITHIN 1 SECOND GRACE PERIOD 1 SECOND ON l.K = m.K "
+              + "%s JOIN %s AS r WITHIN 1 SECOND ON m.K = r.K;",
+          STREAM_1.text(),
+          (joinType == JoinType.OUTER) ? "FULL OUTER" : joinType,
+          STREAM_2.text(),
+          (joinType == JoinType.OUTER) ? "FULL OUTER" : joinType,
+          STREAM_3.text()
+      ));
+    }
+  }
+
+  @Test
+  public void shouldDeprecateStreamStreamJoinsWithNoGraceOnMultiStreamStreamJoinOrder2() {
+    for (final JoinType joinType : JoinType.values()) {
+      checkDeprecatedStatement(String.format(
+          "SELECT * FROM %s AS l "
+              + "%s JOIN %s AS r WITHIN 1 SECOND ON m.K = r.K "
+              + "%s JOIN %s AS m WITHIN 1 SECOND GRACE PERIOD 1 SECOND ON l.K = m.K;",
+          STREAM_1.text(),
+          (joinType == JoinType.OUTER) ? "FULL OUTER" : joinType,
+          STREAM_2.text(),
+          (joinType == JoinType.OUTER) ? "FULL OUTER" : joinType,
+          STREAM_3.text()
+      ));
+    }
+  }
+
+  @Test
+  public void shouldDeprecateStreamStreamJoinsWithNoGraceOnMultiStreamStreamTableJoinOrder1() {
```

Comment:
Should we replicate this test and the next one as "shouldNotWarn" and use GRACE in the stream-stream join?

Reply:
There was one that I renamed to `shouldNotDeprecateStreamStreamJoinsWithGraceOnMultiStreamStreamTableJoinOrder1`. I added another with the name `shouldNotDeprecateStreamStreamJoinsWithGraceOnMultiStreamStreamTableJoinOrder2`.